### PR TITLE
Add global styles

### PR DESCRIPTION
**Pull request by nosolosw**
Part of https://core.trac.wordpress.org/ticket/53175
Follow-up to https://github.com/WordPress/wordpress-develop/pull/1262

## To do

- prepare corresponding Gutenberg PR

## How to test

A theme without theme.json (use the TwentyTwentyOne theme):

- Load the front. Look for a `global-styles-inline-css` embedded stylesheet. The expected result is that none should be found.
- Load the post editor. Search for `--wp--preset--color`. The expected result is that none should be found.

A theme with theme.json support:

- Load the front. Look for a `global-styles-inline-css` embedded stylesheet. The expected result is that there's a stylesheet that contains the CSS generated by the engine. Depending on the theme's theme.json, it should be something along these lines:
```css
body{
  --wp--preset--color--black: #000000;
  /* more CSS Custom Properties for the presets */
}
.has-black-color{ color: #000000 !important; }
.has-black-background-color{ background-color: #000000 !important; }
/* other classes for every preset value  */
```

- Load the post editor. Search for `--wp--preset--color`. The expected result is that there's an embedded stylesheet that contains the CSS Custom Properties.
- Load the post editor. Search for `.has-black-color{ ... }`. The expected result is that there's an embedded stylesheet that contains the classes for the presets as well as any other block styles defined by the theme via `theme.json`.

### CODE REVIEW

**Comment by nosolosw** [comment]
Need to update the unit tests and prepare the corresponding Gutenberg PR, but the code is ready for review.

File: src/wp-includes/script-loader.php [hunk 1]
```diff
@@ -2232,6 +2232,52 @@ function wp_common_block_scripts_and_styles() {
...
 
+/**
+ * Enqueues the global styles defined via theme.json.
+ *
+ * @since 5.8.0
+ *
+ * @return void
+ */
+function wp_enqueue_global_styles() {
+	if ( ! WP_Theme_JSON_Resolver::theme_has_support() ) {
```
Comment (jorgefilipecosta) [inline src/wp-includes/script-loader.php:2243]: In cases where there is link color enabled add_theme_support('experimental-link-color'); we need to output the preset classes, should we have a condition for that here?

File: src/wp-includes/script-loader.php [hunk 2]
```diff
@@ -2232,6 +2232,52 @@ function wp_common_block_scripts_and_styles() {
...
+		}
+	}
+
+	if ( null === $stylesheet ) {
+		$settings   = get_default_block_editor_settings();
+		$theme_json = WP_Theme_JSON_Resolver::get_merged_data( $settings );
+		$stylesheet = $theme_json->get_stylesheet();
+
+		if ( $can_use_cache ) {
+			set_transient( 'global_styles', $stylesheet, MINUTE_IN_SECONDS );
```
Comment (jorgefilipecosta) [inline src/wp-includes/script-loader.php:2268]: Now that this going to be shipped in production would it make sense to have a bigger cache time? cc: @aristath
On another question would it make sense to use `add_action( 'switch_theme', ... );` to clean the cache during a theme switch as we do in https://github.com/WordPress/wordpress-develop/pull/1262/?

File: src/wp-includes/script-loader.php (hunk 1 above)
Comment (nosolosw) [reply src/wp-includes/script-loader.php:2243]: I'm not familiar with prior context on how to deal with this so I'd welcome advice. My thinking is that we should do the following:

- We shouldn't port `experimental-link-color` to WordPress core. Themes that want that should use enable it via `theme.json`.
- In the plugin, I'd expect we remove support for `experimental-link-color` at some point. Probably when the minimum WordPress version for the plugin is 5.8.

Thoughts?

Comment (jorgefilipecosta) [reply src/wp-includes/script-loader.php:2243]: Yes this seems like a plan :

**Review by jorgefilipecosta — APPROVED** [review]
This worked well on my tests and should be ready to merge 👍 
I found some issues setting an empty color palette does not work (it defaults to core) but that's a regression setting an empty color palette is a way to disable the color picker. I guess we should look into these issues separately.

**Comment by jorgefilipecosta** [comment]
Committed in https://github.com/WordPress/wordpress-develop/commit/3d3aa3cd0950f3a44ce87ec8589a40c4460708f4 this PR can be closed.

**Pull request closed without merging**